src/QuickCheckVEngine/RVFI_DII/RVFI.hs

```diff
@@ -73,175 +84,404 @@ type RV_RegIdx = Word8
+                rvfi_mem_rdata = Basement.Types.Word256.Word256 0 0 0 mem_rdata,
+                rvfi_mem_wdata = Basement.Types.Word256.Word256 0 0 0 mem_wdata
+              }
+      }
 
 -- | An otherwise empty halt token for padding
 rvfiEmptyHaltPacket :: RVFI_Packet
-rvfiEmptyHaltPacket = (runGet get (BS.repeat 0)) { rvfi_halt = 1 }
+rvfiEmptyHaltPacket =
+  RVFI_Packet
+    { rvfi_halt = 1,
+      rvfi_valid = 0,
+      rvfi_order = 0,
+      rvfi_insn = 0,
+      rvfi_trap = 0,
+      rvfi_intr = 0,
+      rvfi_mode = Nothing,
+      rvfi_ixl = Nothing,
+      rvfi_pc_rdata = 0,
+      rvfi_pc_wdata = 0,
+      rvfi_int_data = Nothing,
+      rvfi_mem_data = Nothing
+    }
+
+prvString :: Maybe RV_PrivMode -> String
+prvString Nothing = "PRV_?"
+prvString (Just x) = show x
 
 instance Show RVFI_Packet where
   show tok
-    | rvfiIsHalt tok = "halt token"
-    | otherwise = printf "Trap: %5s, PCWD: 0x%016x, RD: %02d, RWD: 0x%016x, MA: 0x%016x, MWD: 0x%016x, MWM: 0b%08b, I: 0x%016x (%s)"
-                  (show $ rvfi_trap tok /= 0) -- Trap
-                  (rvfi_pc_wdata tok)    -- PCWD
-                  (rvfi_rd_addr tok)     -- RD
-                  (rvfi_rd_wdata tok)    -- RWD
-                  (rvfi_mem_addr tok)    -- MA
-                  (rvfi_mem_wdata tok)   -- MWD
-                  (rvfi_mem_wmask tok)   -- MWM
-                  (rvfi_insn tok) (pretty (toInteger (rvfi_insn tok))) -- Inst
+    | rvfiIsHalt tok = printf "halt token v%d" (rvfiHaltVersion tok)
+    | otherwise =
+      printf
+        "Trap: %5s, PCWD: 0x%016x, RD: %02d, RWD: 0x%016x, MA: 0x%016x, MWD: 0x%016x, MWM: 0b%08b, I: 0x%016x %s (%s)"
+        (show $ rvfi_trap tok /= 0) -- Trap
+        (rvfi_pc_wdata tok) -- PCWD
+        (rvfi_rd_addr intData) -- RD
+        (rvfi_rd_wdata intData) -- RWD
+        (rvfi_mem_addr memData) -- MA
+        (toNatural (rvfi_mem_wdata memData)) -- MWD
+        (rvfi_mem_wmask memData) -- MWM
+        (rvfi_insn tok)
+        (prvString (rvfi_mode tok))
+        (pretty (toInteger (rvfi_insn tok))) -- Inst
+    where
+      intData = fromMaybe rvfiEmptyIntData (rvfi_int_data tok)
+      memData = fromMaybe rvfiEmptyMemData (rvfi_mem_data tok)
 
 -- | Return 'True' for halt 'RVFI_Packet's
 rvfiIsHalt :: RVFI_Packet -> Bool
 rvfiIsHalt x = rvfi_halt x /= 0
 
+-- | Return the trace version for a halt packet (using the previously-unused bits)
+rvfiHaltVersion :: RVFI_Packet -> Word8
+rvfiHaltVersion x = 1 + shiftR (rvfi_halt x) 1
+
 -- | Return 'True' for trap 'RVFI_Packet's
 rvfiIsTrap :: RVFI_Packet -> Bool
 rvfiIsTrap x = rvfi_trap x /= 0
 
+-- | Since version 1 trace format implementations don't report the current privilege
+-- | level we can't fail the comparison if one implementation reports a privilege level
+-- | and the other one does
+privLevelSame :: (Maybe RV_PrivMode) -> (Maybe RV_PrivMode) -> Bool
+privLevelSame _ Nothing = True
+privLevelSame Nothing _ = True
+privLevelSame x y = (x == y)
```

Comment:
```suggestion
privLevelSame :: (Maybe RV_PrivMode) -> (Maybe RV_PrivMode) -> (Maybe Bool)
privLevelSame = mzipWith (==)
```
and then let callers unpack it (which is also nice, because `privLevelSame` notionally doesn't have a known value for a Nothing input), or if you really want to return a plain Bool here:
```suggestion
privLevelSame :: (Maybe RV_PrivMode) -> (Maybe RV_PrivMode) -> Bool
privLevelSame = fromMaybe False $ mzipWith (==)
```
NB: mzipWith is in Control.Monad.Zip

Reply:
Although either way maybe you don't need the function...

Reply:
That looks a lot better. Thanks. And it should be possible to reuse that for other types too.